src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -167,7 +167,35 @@ internal partial class DiagnosticIncrementalAnalyzer
                     {
                         Logger.Log(FunctionId.Diagnostics_ProjectDiagnostic, p => $"FSA off ({p.FilePath ?? p.Name})", project);
 
-                        return new ProjectAnalysisData(project.Id, VersionStamp.Default, existingData.Result, ImmutableDictionary<DiagnosticAnalyzer, DiagnosticAnalysisResult>.Empty);
+                        // If we are producing document diagnostics for some other document in this project, we still want to show
+                        // certain project-level diagnostics that would cause file-level diagnostics to be broken. We will only do this though if
+                        // some file that's open is depending on this project though -- that way we're going to only be analyzing projects
+                        // that have already had compilations produced for.
```

Comment:
Tagging @CyrusNajmabadi. Cyrus and me have been discussing about how to reduce the number of compilations getting produced via background diagnostic computation by avoiding analyzing all open files for compiler diagnostics. We are hoping to soon introduce a new background analysis mode that would only refresh compiler diagnostics for visible documents (we will dogfood + design this further to make sure it doesn't break user experience).

However, your change below would mean that we will always produces compilations for projects which have at least one open document or are being referenced by a project which has one open document, which will nullify any of the above mentioned perf optimizations as per the current implementation below. `GetProjectAnalysisDataAsync` does get invoked by the solution crawler at a longer delay though, so we may be fine in terms of responsiveness hit by doing this additional compute, but something that we certainly need to keep an eye on in terms of performance.

Reply:
@mavasan, @CyrusNajmabadi: I'd have no problem with this code here then also changing to "visible" documents as well as a part of that change if you wanted to go ahead with that. Honestly if there's a function I should be calling that is "FSA is off, what documents will be analyzed?" that I should be calling here now, happy to switch it.

Reply:
Also this made me realize that here:

https://github.com/dotnet/roslyn/blob/4814028f1e138ed17a7d05f76d8532a0c4ed5687/src/Features/Core/Portable/Workspace/BackgroundCompiler.cs#L101

We are asking for compilations for all open documents's projects there, and that'll be more aggressive than here anyways. Going to look at making an experiment to either change that to visible files, or just get rid of the background compiler entirely.

Reply:
Sounds fine, we can take this up as a follow-up work item.

Reply:
I very frequently open a group of related files, make a change, and watch to see the error list update for all the open documents. Changing the compiler diagnostics to only run on visible documents would be severely disruptive to what seems like a common workflow.

Reply:
> Changing the compiler diagnostics to only run on visible documents would be severely disruptive to what seems like a common workflow

@sharwell Our current plan is to have the default still be "Open Files" for compiler diagnostics. We will add the new option for "Visible Files", have folks on the IDE team dogfood that mode and determine if there is value in keeping in and/or making it the default mode. I was planning to bring this up for discussion sometime this week or early next within our team.